doc/design/design_support_arbitrary_select.md

```diff
@@ -0,0 +1,57 @@
+# Support Arbitrary Select Statements
+
+SQLFlow extends the SQL syntax to enable model training and inference. The syntax extension should be easy to learn and integrate well with the existing SQL syntax. While the standard SQL syntax supports complicated select statements, SQLFlow should support training/predict on these statements via appending a TRAIN/PREDICT clause right after them.
+
+With the support of arbitrary select statements, a user can integrate AI into his/her database with ease.
+
+## Overview
+
+Given a long and nested select inside an extended SQL statement like
+
+```sql
+-- select clause
+SELECT c1, c2, id, class FROM (SELECT * FROM my_table) a
+-- train clause
+TRAIN DNNClassifier
+WITH n_class = 3
+COLUMN c1, c2, EMBEDDING(id)
+LABEL class
+INTO my_dnn_model;
+```
+
+SQLFlow does the following steps.
+
+1. SQLFlow splits the extended SQL statement into its select clause and its train clause. In the above example, the select clause is `SELECT ... a` and the train clause is `TRAIN DNNClassifier ... INTO my_dnn_model`. For the train clause, we check the syntax by our parser at `pkg/sql/parser.go`.
+
+1. SQLFlow verifies the column in the train clause.
+    1. SQLFlow executes the select clause and retrieves the column names and column types of the result. For example, the result of `SELECT ... a` has four columns with names `c1`, `c2`, `id`, and `class`. `c1`, and `c2` are of float types. And `id` and `class` are of integer types.
```

Comment:
Is it vital to use APIs like `describe table` to retrieve column names and types? Running the arbitrary select statement maybe resource consuming.

Reply:
Just to clarify, you are referring to create a temporary table using `CREATE TABLE {.TempTable} AS ({.SelectClause});`, then use `describe table` to retrieve column names and types. By doing so we can avoid running the arbitrary select in the later feature derivation stage and training stage.

If the user has the write permission to the DB, I do believe this method can have performance improvement. However, I would like to save these details until we find performance is the major issue.

Reply:
> However, I would like to save these details until we find performance is the major issue.

Sounds good.